Number of samples: `17,570`

Number of features: `5`

Target: `total_sales(mil)` = total sales in millions

Description: video game dataset with metadata (`console`, `genre`, `publisher`, `developer`, `release date`), a critic score, and sales data.

Goal: predict `total_sales(mil)` from `console`, `genre`, `publisher`, `developer`, and `release_year`.
We do not use regional sales to avoid data leakage.
We also remove the critic score because about 90% of the values are missing.

Industry interest: estimate commercial potential and understand which factors are linked to sales.


## Data Preparation


In [8]:
import numpy as np
import pandas as pd


def convert_sales_in_number(df):
    df = df.copy()
    df.columns = df.columns.str.strip()

    sales_columns = [
        "na_sales(mil)",
        "jp_sales(mil)",
        "pal_sales(mil)",
        "other_sales(mil)",
        "total_sales(mil)",
    ]

    for col in sales_columns:
        if col in df.columns:
            values = df[col].astype(str).str.strip()
            values = values.str.replace("\u00A0", "", regex=False)
            values = values.str.replace(r"[^0-9\-\.,]", "", regex=True)
            values = values.str.replace(",", ".", regex=False)
            df[col] = pd.to_numeric(values, errors="coerce")

    return df


df = pd.read_csv("regression_datasets/VideoGames_Sales.csv", sep=";")
df.columns = df.columns.str.strip()
df = convert_sales_in_number(df)

target = "total_sales(mil)"
df = df.dropna(subset=[target])
df = df[df[target] >= 0].copy()

feature_cols = ["console", "genre", "publisher", "developer", "release_date"]
X = df[feature_cols].copy()
X["release_date"] = pd.to_datetime(X["release_date"], errors="coerce", dayfirst=True)
X["release_year"] = X["release_date"].dt.year
X = X.drop(columns=["release_date"])

y = np.log1p(df[target])

print("Samples:", len(df))
print("X shape:", X.shape)
print("Target:", target, "(log1p transform applied)")
print("Missing values per column:\n", X.isna().sum())
X.head()


Samples: 17570
X shape: (17570, 5)
Target: total_sales(mil) (log1p transform applied)
Missing values per column:
 console          0
genre            0
publisher        0
developer        2
release_year    62
dtype: int64


,console,genre,publisher,developer,release_year
0,PS3,Action,Rockstar Games,Rockstar North,2013.0
1,PS4,Action,Rockstar Games,Rockstar North,2014.0
2,PS2,Action,Rockstar Games,Rockstar North,2002.0
3,X360,Action,Rockstar Games,Rockstar North,2013.0
4,PS4,Shooter,Activision,Treyarch,2015.0


## Model Comparison

We split the data into training and test sets and compare several regression models.
Categorical variables are one-hot encoded, missing values are imputed, and a `Dummy` model is used as a baseline.


In [9]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.compose import ColumnTransformer
from sklearn.dummy import DummyRegressor
from sklearn.ensemble import HistGradientBoostingRegressor, RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder


class ToDense(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.is_fitted_ = True
        return self

    def transform(self, X):
        return X.toarray() if hasattr(X, "toarray") else X


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

num_cols = ["release_year"]
cat_cols = ["console", "genre", "publisher", "developer"]


def build_preprocessor(dense=False):
    preprocess_sparse = ColumnTransformer(
        transformers=[
            ("num", SimpleImputer(strategy="median"), num_cols),
            (
                "cat",
                Pipeline(
                    steps=[
                        ("imputer", SimpleImputer(strategy="most_frequent")),
                        ("onehot", OneHotEncoder(handle_unknown="ignore")),
                    ]
                ),
                cat_cols,
            ),
        ]
    )

    if dense:
        return Pipeline(
            steps=[
                ("prep", preprocess_sparse),
                ("todense", ToDense()),
            ]
        )

    return preprocess_sparse


def evaluate(model, name):
    prep = build_preprocessor(dense=(name == "hgb"))
    pipe = Pipeline([("prep", prep), ("model", model)])
    pipe.fit(X_train, y_train)
    pred = pipe.predict(X_test)

    mae = mean_absolute_error(y_test, pred)
    mse = mean_squared_error(y_test, pred)
    rmse = mse ** 0.5
    r2 = r2_score(y_test, pred)

    print(f"{name:>6} | MAE(log): {mae:.4f} | RMSE(log): {rmse:.4f} | R2: {r2:.4f}")
    return pipe


models = [
    ("dummy", DummyRegressor(strategy="mean")),
    ("ridge", Ridge()),
    ("rf", RandomForestRegressor(n_estimators=300, random_state=42, n_jobs=-1)),
    ("hgb", HistGradientBoostingRegressor(random_state=42)),
]

trained = {}
for name, model in models:
    trained[name] = evaluate(model, name)


 dummy | MAE(log): 0.2147 | RMSE(log): 0.3158 | R2: -0.0000
 ridge | MAE(log): 0.1604 | RMSE(log): 0.2408 | R2: 0.4185
    rf | MAE(log): 0.1400 | RMSE(log): 0.2369 | R2: 0.4373
   hgb | MAE(log): 0.1567 | RMSE(log): 0.2449 | R2: 0.3985


### Observed results on the log scale

The notebook gives the following scores:

```text
 dummy | MAE(log): 0.2147 | RMSE(log): 0.3158 | R2: -0.0000
 ridge | MAE(log): 0.1604 | RMSE(log): 0.2408 | R2: 0.4185
    rf | MAE(log): 0.1400 | RMSE(log): 0.2369 | R2: 0.4373
   hgb | MAE(log): 0.1567 | RMSE(log): 0.2449 | R2: 0.3985
```

Main points:
- the baseline model explains almost nothing, which is expected;
- `Ridge` clearly improves the prediction;
- `RandomForestRegressor` is the best model on this split;
- `HistGradientBoostingRegressor` is still good, but a little below `Ridge` and `RandomForest` here.


In [10]:
def eval_original_units(pipe, label):
    pred_log = pipe.predict(X_test)
    y_pred = np.expm1(pred_log)
    y_true = np.expm1(y_test)

    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = mse ** 0.5

    print(f"{label:>6} | MAE: {mae:.3f} million | RMSE: {rmse:.3f} million")


eval_original_units(trained["ridge"], "ridge")
eval_original_units(trained["rf"], "rf")
eval_original_units(trained["hgb"], "hgb")


 ridge | MAE: 0.273 million | RMSE: 0.691 million
    rf | MAE: 0.248 million | RMSE: 0.627 million
   hgb | MAE: 0.272 million | RMSE: 0.701 million


### Interpretation in original sales units

Results in the original scale:
- `RandomForest`: MAE about `0.248` million, RMSE about `0.627` million
- `Ridge`: MAE about `0.273` million, RMSE about `0.691` million
- `HGB`: MAE about `0.272` million, RMSE about `0.701` million

So the average absolute error is around `0.25` to `0.27` million copies, or about `250,000` to `270,000` units. The RMSE is higher because it gives more weight to large errors on blockbuster games.


### Largest Prediction Errors


In [11]:
rf_pred_log = trained["rf"].predict(X_test)
rf_pred = np.expm1(rf_pred_log)
y_true_original = np.expm1(y_test)

error_analysis = X_test.reset_index(drop=True).copy()
error_analysis["true_sales_million"] = y_true_original.reset_index(drop=True)
error_analysis["predicted_sales_million"] = pd.Series(rf_pred)
error_analysis["absolute_error_million"] = (
    error_analysis["true_sales_million"] - error_analysis["predicted_sales_million"]
).abs()

top_10_errors = error_analysis.sort_values("absolute_error_million", ascending=False).head(10)
display(top_10_errors[[
    "console",
    "genre",
    "publisher",
    "developer",
    "release_year",
    "true_sales_million",
    "predicted_sales_million",
    "absolute_error_million",
]])


,console,genre,publisher,developer,release_year,true_sales_million,predicted_sales_million,absolute_error_million
2934,X360,Shooter,Activision,Infinity Ward,2011.0,14.82,1.986532,12.833468
118,PS3,Action,Rockstar Games,Rockstar North,2008.0,10.57,1.626322,8.943678
2915,PS3,Shooter,Activision,Treyarch,2010.0,12.67,4.650060,8.019940
623,PS3,Action,Rockstar Games,Rockstar San Diego,2011.0,6.57,0.359885,6.210115
2246,PS4,Role-Playing,Bethesda Softworks,Bethesda Game Studios,2015.0,8.48,2.380795,6.099205
3251,X360,Role-Playing,Bethesda Softworks,Bethesda Game Studios,2011.0,8.88,2.848520,6.031480
1351,PS4,Misc,Sony Computer Entertainment,Mojang,2014.0,6.33,0.431244,5.898756
2535,X360,Shooter,Activision,Treyarch,2012.0,13.86,8.546436,5.313564
2495,PSP,Action,Rockstar Games,Rockstar Leeds,2005.0,7.72,2.835144,4.884856
1950,PS3,Sports,EA Sports,EA Canada,2010.0,5.08,0.675281,4.404719


## Discussion

The results are clear:
- the `DummyRegressor` confirms the baseline, with `R2` close to zero;
- the metadata already contain useful information, because `Ridge`, `RandomForest`, and `HGB` all do much better than the baseline;
- the best result is obtained with `RandomForestRegressor`, with `R2` around `0.44`;
- this means the model explains about `44%` of the variance on the test set, which is useful but still limited.

This level of performance is realistic. Video game sales depend on many factors that are not in the dataset: marketing, franchise strength, release timing, review quality, competition, and platform strategy. So the model is useful for a rough estimate, but not for very precise forecasting.

The table of the largest errors confirms the same idea: the model works reasonably well on ordinary games, but it often misses exceptional hits. The biggest errors are usually linked to very high-selling games, so the model tends to underestimate blockbuster titles.


## Conclusion

We trained several regression models to predict `total_sales(mil)` from metadata only. The target was transformed with `log1p` because sales are very uneven.

Among the tested models, `RandomForestRegressor` is the best choice on this split. It gives the best `R2` and the lowest error on the log scale. In original units, the average error is a few hundred thousand copies.

So this approach is useful to estimate an order of magnitude, but it is not enough for precise business forecasting. The main limits are the missing explanatory variables and the fact that only about `17,570` samples remain after removing rows with missing target values.
